examples/ws_subscription.rs

```diff
@@ -44,33 +50,51 @@ async fn main() -> anyhow::Result<()> {
 	let addr = run_server().await?;
 	let url = format!("ws://{}", addr);
 
-	let client = WsClientBuilder::default().build(&url).await?;
-	let mut subscribe_hello: Subscription<String> =
-		client.subscribe("subscribe_hello", rpc_params![], "unsubscribe_hello").await?;
+	let client1 = WsClientBuilder::default().build(&url).await?;
+	let client2 = WsClientBuilder::default().build(&url).await?;
+	let sub1: Subscription<i32> = client1.subscribe("subscribe_hello", rpc_params![], "unsubscribe_hello").await?;
+	let sub2: Subscription<i32> = client2.subscribe("subscribe_hello", rpc_params![], "unsubscribe_hello").await?;
 
-	let mut i = 0;
-	while i <= NUM_SUBSCRIPTION_RESPONSES {
-		let r = subscribe_hello.next().await;
-		tracing::info!("received {:?}", r);
-		i += 1;
-	}
+	let fut1 = sub1.take(NUM_SUBSCRIPTION_RESPONSES).for_each(|r| async move { tracing::info!("sub1 rx: {:?}", r) });
+	let fut2 = sub2.take(NUM_SUBSCRIPTION_RESPONSES).for_each(|r| async move { tracing::info!("sub2 rx: {:?}", r) });
+
+	future::join(fut1, fut2).await;
 
 	Ok(())
 }
 
 async fn run_server() -> anyhow::Result<SocketAddr> {
 	let server = WsServerBuilder::default().build("127.0.0.1:0").await?;
 	let mut module = RpcModule::new(());
-	module.register_subscription("subscribe_hello", "s_hello", "unsubscribe_hello", |_, mut sink, _| {
-		std::thread::spawn(move || loop {
-			if let Err(Error::SubscriptionClosed(_)) = sink.send(&"hello my friend") {
-				return;
-			}
-			std::thread::sleep(std::time::Duration::from_secs(1));
+	let (tx, _) = broadcast::channel(1024);
+
+	tokio::spawn(produce_items(tx.clone()));
+
+	module.register_subscription("subscribe_hello", "s_hello", "unsubscribe_hello", move |_, sink, _| {
+		let rx = tx.subscribe();
+
+		// Convert stream from `Item = Result<T: Serialize, Error>` to `Item = T::Serialize`.
+		let stream =
+			BroadcastStream::new(rx).take_while(|r| future::ready(r.is_ok())).filter_map(|r| future::ready(r.ok()));
```

Comment:
This is a bit unfortunate, a lot of line-noise just to massage the stream to do the right thing. Any chance it can be simpler?

Reply:
yeah, we need to change to API for stream to take `impl Stream<Item = Result<T::Serialize, Error>>` instead of `impl Stream<Item = T: Serialize` then this goes away but other streams such `mpsc::Receiver` need to wrap it in a Result which more flexible.

Reply:
switched to the `async-broadcast` the only reason for `tokio::sync::Broadcast` to fail is if the receiver can't keep up with sender. 

similar to what we do in the `ws client` but for readability let's use async-broadcast